airflow/opt/providers/etna/etna/operators/rollup_xcom_operator.py

```diff
@@ -0,0 +1,106 @@
+from datetime import datetime
+from functools import reduce
+from typing import Any, Optional, Callable, Iterable, Tuple, Dict
+
+from airflow.models import BaseOperator, XCom, TaskInstance
+from airflow.models.xcom_arg import XComArg
+from airflow.utils.session import provide_session
+from airflow.utils.state import State
+
+from etna.etls.etl_task_batching import get_batch_range
+from sqlalchemy.orm import Session
+
+from etna.metrics.rollup_metrics import Rollup
+
+
+class RollupState(Rollup):
+    last_execution_date: datetime
+    result: Rollup
+
+    def __init__(self, last_execution_date: datetime, result: Rollup):
+        self.result = result
+        self.last_execution_date = last_execution_date
+
+    def measure(self) -> Iterable[Tuple[Dict[str, str], int]]:
+        return self.result.measure()
+
+class rollup:
+    concat: Callable[[Rollup, Rollup], Rollup]
+
+    def __init__(self, concat: Callable[[Rollup, Rollup], Rollup]):
+        self.concat = concat
+
+    def __call__(self, arg: XComArg):
+        op = RollupXcomOperator(concat=self.concat, arg=arg)
+        arg >> op
+        return op
+
+class RollupXcomOperator(BaseOperator):
+    concat: Callable[[Any, Any], Any]
+    target_dag_id: str
+    target_task_id: str
+
+    def __init__(self,
+                 concat: Callable[[Rollup, Rollup], Rollup],
+                 arg: Optional[XComArg] = None,
+                 target_dag_id: Optional[str] = None,
+                 target_task_id: Optional[str] = None,
+                 task_id: Optional[str] = None, **kwargs
+                 ):
+        if arg is not None:
+            target_dag_id = target_dag_id or arg.operator.dag_id
+            target_task_id = target_task_id or arg.operator.task_id
+
+        super().__init__(task_id=task_id or f"rollup_{target_dag_id}_{target_task_id}", **kwargs)
+        self.concat = concat
+
+        self.target_dag_id = target_dag_id
+        self.target_task_id = target_task_id
+
+    @provide_session
+    def execute(self, context: Any, session: Session = None):
+        start, stop = get_batch_range(context)
+
+        result: Optional[RollupState] = None
+
+        ti: TaskInstance = context['ti']
+        previous: Optional[TaskInstance] = ti.get_previous_ti(state=State.SUCCESS)
```

Comment:
Essentially, a rollup is a task using an operator `concat` to summarize a collection of XCom results of other tasks.

Idea is that eg a magma metric generator takes samples of the state of magma records which are 'rolled up' into a larger picture assessment of the current completeness of a data set.

Reply:
How often do the rollups run? Does this mean that they are like Prometheus metrics (since above you note that it takes the last XCom result of a given task) -- sampling of other task results, not a comprehensive summation?

Reply:
Rollups are just standard DAG tasks, so they run as often as you configure them.

They can be a comprehensive summation -- it entirely depends on how you implement the concat operation.  The operator itself attempts to roll up all xcom results of another dag.  How you sum that is dependent on the rollup you are building.  So for completion metrics, this will be a full summation.